File: Algo/Leetcode/121BestTimeToBuyAndSellStock.cpp
// dp. Use two array to store min and max
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        if(n<=1) return 0;
        vector<int> min_v(n);
        vector<int> max_v(n);
        min_v[0] = prices[0];
        max_v[0] = prices[n-1];
        for(int i=1; i<n; i++){
            min_v[i] = min(min_v[i-1], prices[i]);
            max_v[i] = max(max_v[i-1], prices[n-i-1]);
        }

        int max_profit = 0;
        for(int i=0; i<n; i++){
            max_profit = max(max_profit, max_v[n-i-1] - min_v[i]);
        }
        return max_profit;
    }
};

// Just store the lowest price and get largest profit
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        if(n<=1) return 0;
        int min_v = INT_MAX;
        int max_profit = 0;
        for(int i=0; i<n; i++){
            min_v = min(min_v, prices[i]);
            max_profit = max(max_profit, prices[i] - min_v);
        }

        return max_profit;
    }
};

File: Algo/Leetcode/121BestTimeToBuyAndSellStock.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        max_r = [prices[-1]]
        for n in prices[:-1][::-1]:
            max_r.append(max(max_r[-1], n))
        max_r = max_r[::-1]

        diff = [max_r[i] - prices[i] for i in range(len(prices))]
        return max(max(diff), 0)


# just stop max profit
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        max_profit = 0
        min_v = prices[0]
        for p in prices[1:]:
            min_v = min(min_v, p)
            max_profit = max(max_profit, p - min_v)

        return max_profit